mono/mini/aot-mangler.c

```diff
@@ -0,0 +1,1244 @@
+/**
+ * \file
+ * Aot Mangler Functions
+ *
+ * Authors:
+ *   Alexander Kyte (alkyte@microsoft.com)
+ *
+ * Copyright 2017 Microsoft, Inc (http://www.microsoft.com)
+ * Licensed under the MIT license. See LICENSE file in the project root for full license information.
+ */
+
+#include <mono/utils/json.h>
+#include "config.h"
+#include "aot-compiler.h"
+
+/*
+ * Mono Method Mangler Format:
+ *
+ * It is often useful to have a unique string associated with
+ * each method. The complications with using this to describe so
+ * rich a type system as C# is the issue of transitive type references.
+ *
+ * It is necessary to bake in information about the referenced types which 
+ * becomes quite verbose when they nest. Clarity around the depth of the tree
+ * at a given point in the string is important. We elect to use escaped brackets.
+ *
+ * Without using special characters, this is pretty difficult to do without
+ * risking a clash with user-provided names. For this reason we encode the mangled
+ * name in a format that is later scrubbed of these characters. This necessitates that we
+ * make our emitted replacements unique, unable to clash with user-provided names. In our case
+ * this is done by escaping the separator character we insert, the underscore.
+ *
+ * This makes for long, difficult-to-read strings with lots of structural nesting.
```

Comment:
It would be nice if the description included a grammar/description of the encoding itself.